은행 창구 관리앱 [STEP 1] - d.o & joo

doshkor opened the pull request:
안녕하세요🙋‍♂️ @wonhee009

이번 프로젝트 `은행 창구 관리앱`을 진행하게 된 D.o, Joo 입니다. 리뷰어로 라자냐를 만나게 되어서 반가워요. 
STEP1에서는 자료구조 Queue를 구현하는 것으로 간단?했어요. 2주가 아닌 3주 동안의 좋은 리뷰 부탁드려요. 감사합니다

## **Step 1 : 큐 타입 구현**

- [x]  Queue 타입 구현을 위한 Linked-list 타입을 직접 구현합니다
- [x]  Queue 타입은 다양한 타입의 데이터를 취급할 수 있도록 스위프트의 Generics 기능을 활용합니다
- [x]  Queue 타입이 예상한대로 동작하는지 Unit Test를 통해 동작을 검증합니다
- [x]  Queue에 구현해야 할 필수 기능 : `Enqueue`, `Dequeue`, `Clear`, `Peek`, `isEmpty`

### **스텝 핵심 경험**

- Linked-list 자료구조의 이해 및 구현
- Queue 자료구조의 이해 및 구현
- Generics 개념이해 및 적용

# 질문

1️⃣ `isEmpty 와 peek` 를 `메서드와 계산속성` 중 어떤 것으로 구현해야 할지 궁금합니다.!

- LinkedList 에서 isEmpty 와 peek 는 다른 속성의 상태를 변경하지 않기 때문에 계산속성으로 구현하는 방식도 구현해 보았어요. (read-only)
- Queue<T> 에서도  같은 방법으로 계산속성으로 나타내어도 괜찮을까요? 메서드가 아닌 계산속성을 이용하는데 어색한 부분이 있지는 않을지 궁금해서 질문 드립니다

2️⃣ `Queue`의 메서드는 단순히 `LinkedList` 타입의 메스드를 불러와서 동작시켰습니다

LinkedList의 메서드마다 테스트 거쳐서 안정성이 확보되었다고 생각이 들었습니다. Queue의 모든 메서드는 LinkedList의 메서드를 사용하기에 UnitTest를 하는 것이 중복된 테스트이며 불필요한 검증이라 생각이 들었습니다. 테스트를 진행하지 않는 게 맞는 건지 궁금합니다

altokeys committed:
first commit

ohdair committed:
docs: .gitignore 수정

ohdair committed:
feat: LinkedList 타입 추가

- Node 타입 추가
- Node를 추가하는 append 메서드 추가
- Unit Test로 append 함수 동작을 확인

altokeys committed:
feat: removeFirst 메서드 추가 및 Unit Test 동작 확인

ohdair committed:
feat: LinkedList의 isEmpty, clear 메서드 추가

- UnitTest를 통한 테스트

altokeys committed:
feat: Queue 생성 및 메서드 추가

ohdair committed:
refactor: 내부 변수를 은닉

- isEmpty 연산 프로퍼티로 변경
- list의 프로퍼티 head 직접 접근을 막고 연산 프로퍼티로 값을 제공
- 코드 변경에 따른 Test 변경

ohdair committed:
refactor: mutating 키워드 삭제

wonhee009 commented on BankManagerConsoleApp/BankManagerConsoleApp/Queue.swift:
```diff
@@ -0,0 +1,32 @@
+        return list.removeFirst()
+    }
+    
+    func clear() {
```
`clear`보다는 `removeAll`은 어떨까요? [apple array](https://developer.apple.com/documentation/swift/array)에서도 removeAll이기도 하구요.

wonhee009 commented on BankManagerConsoleApp/BankManagerConsoleAppTests/BankManagerConsoleAppTests.swift:
```diff
@@ -0,0 +1,76 @@
+        XCTAssertTrue(sut.isEmpty)
+    }
+
+    func test_append_입력_3에_대해서_추가되었는지_확인한다() throws {
+        let input = 3
+        sut.append(input)
+        XCTAssertFalse(sut.isEmpty)
+    }
```
3이 제대로 들어갔는지 확인하는 부분은 `isEmpty`를 이용하는 것보다는 `peek`을 이용하는게 더 정확하지 않을까요?

wonhee009 reviewed (changes requested):
수고하셨습니다 :)
코멘트 남겨두었으니 확인부탁드려요.

- 큐도 마찬가지로 상태를 변경하지 않으니 computed property로 구현해도 좋을거 같습니다.
- 테스트 코드를 작성하는 이유는 유지보수 하면서 내부 로직이 변경되어도 항상 동일한 결과값을 보장하기 위한 기능도 있다고 생각하는데요. 현재는 Queue가 LinkedList의 메서드를 그대로 가져다 쓰고 있어 테스트 코드가 중복 혹은 불필요하다고 생각할 수 있을거 같습니다. 다만, Queue를 구현하는 방식이 LinkedList일 필요는 없고, 이후에 유지보수를 하다 Array 등 다른 방식으로 Queue를 구현하게 된다면 이전과 동일한 동작임을 보장 혹은 테스트할 수 있는 코드가 필요하지 않을까요?

ohdair committed:
refactor: LinkedLIst clear 메서드 이름 removeAll로 변경

- apple Array에서도 사용되는 removeAll로 동일하게 변경
- Test에서 추가를 확인하는 검증에서 isEmpty가 아닌 peek으로 확인하도록 변경
